solr/contrib/gcs-repository/src/java/org/apache/solr/gcs/GCSBackupRepository.java

```diff
@@ -0,0 +1,466 @@
+            return PathType.DIRECTORY;
+
+        return PathType.FILE;
+    }
+
+    private String toBlobName(URI path) {
+        return path.toString();
+    }
+
+    @Override
+    public String[] listAll(URI path) throws IOException {
+        String blobName = path.toString();
+        if (!blobName.endsWith("/"))
+            blobName += "/";
+
+        final String pathStr = blobName;
+        final LinkedList<String> result = new LinkedList<>();
+        storage.list(
+                bucketName,
+                Storage.BlobListOption.currentDirectory(),
+                Storage.BlobListOption.prefix(pathStr),
+                Storage.BlobListOption.fields())
+        .iterateAll().forEach(
+                blob -> {
+                    assert blob.getName().startsWith(pathStr);
+                    final String suffixName = blob.getName().substring(pathStr.length());
+                    if (!suffixName.isEmpty()) {
+                        // Remove trailing '/' if present
+                        if (suffixName.endsWith("/")) {
+                            result.add(suffixName.substring(0, suffixName.length() - 1));
+                        } else {
+                            result.add(suffixName);
+                        }
+                    }
+                });
+
+        return result.toArray(new String[0]);
+    }
+
+    @Override
+    public IndexInput openInput(URI dirPath, String fileName, IOContext ctx) throws IOException {
+        return openInput(dirPath, fileName, ctx, 2 * 1024 * 1024);
+    }
+
+    private IndexInput openInput(URI dirPath, String fileName, IOContext ctx, int bufferSize) {
+        String blobName = dirPath.toString();
+        if (!blobName.endsWith("/")) {
+            blobName += "/";
+        }
+        blobName += fileName;
+
+        final BlobId blobId = BlobId.of(bucketName, blobName);
+        final Blob blob = storage.get(blobId, Storage.BlobGetOption.fields(Storage.BlobField.SIZE));
+        final ReadChannel readChannel = blob.reader();
+        readChannel.setChunkSize(bufferSize);
+
+        return new BufferedIndexInput(blobName, bufferSize) {
+
+            @Override
+            public long length() {
+                return blob.getSize();
+            }
+
+            @Override
+            protected void readInternal(ByteBuffer b) throws IOException {
+                readChannel.read(b);
+            }
+
+            @Override
+            protected void seekInternal(long pos) throws IOException {
+                readChannel.seek(pos);
+            }
+
+            @Override
+            public void close() throws IOException {
+                readChannel.close();
+            }
+        };
+    }
+
+    @Override
+    public OutputStream createOutput(URI path) throws IOException {
+        final BlobInfo blobInfo = BlobInfo.newBuilder(bucketName, toBlobName(path)).build();
+        final Storage.BlobWriteOption[] writeOptions = new Storage.BlobWriteOption[0];
```

Comment:
Should we call storage.writer() without the second arg, or should we have a constant NO_WRITE_OPTIONS?

Reply:
Good idea - I ended up creating a constant called `DEFAULT_WRITE_OPTIONS`.  The value is still an empty array, but having the constant lets users subclass and override if desired.